### PR TITLE
Added eic driver for sam4l

### PR DESCRIPTION
**Pull request by jiashuoz**
### Pull Request Overview

This pull request adds the hardware driver for the external interrupt controller module in sam4l.
Mainly, two files were created and added: chips/sam4l/src/eic.rs and kernel/src/hil/eic.rs.

### Testing Strategy

This pull request was tested by enabling an external interrupt on the user push button (pin PC[24]). Each button press will print a message using debug!().

### TODO or Help Wanted

This pull request still needs further testing on the difference between asynchronous mode and synchronous operations. This pull request still needs further testing on the difference between the level-triggered interrupts and the edge-triggered interrupts. 

We still need to figure out how NMI works with EIC.

### Documentation Updated

- [x] No update is needed at this stage.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: kernel/src/hil/eic.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+//! Interface for external interrupt controller.
```
Comment (bradjc) [inline kernel/src/hil/eic.rs]: Having some more description here that provides future implementors some intuition on _when_ to use this HIL would be helpful.

File: kernel/src/hil/eic.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+/// Enum for selecting syn/asyn mode
+pub enum SynchronizationMode {
+    Synchronous,
+    Asynchronous,
+}
+
+/// Interface for EIC.
+pub trait ExternalInterruptController {
```
Comment (bradjc) [inline kernel/src/hil/eic.rs]: Would this make sense to be called `DirectInterrupt`? Or perhaps External Interrupt is a more common name than I imagine.

File: kernel/src/hil/eic.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    /// Enables EIC module
+    fn enable(&self);
+
+    /// Disables EIC module
+    fn disable(&self);
+
+    /// Enables external interrupt on line_num
+    /// In asychronous mode, all edge interrupts will be interpreted as level interrupts and the filter is disabled.
+    fn line_enable(&self, line_num: usize);
```
Comment (bradjc) [inline kernel/src/hil/eic.rs]: Adding a type to this trait and making `line_num` of that type (see the ADC and analog comparator HILs for examples) would be more consistent.

File: kernel/src/hil/eic.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    fn line_disable(&self, line_num: usize);
+
+    /// Configure external interrupt on line_num
+    fn line_configure(
+        &self,
+        line_num: usize,
+        int_mode: InterruptMode,
+        filter: FilterMode,
+        syn_mode: SynchronizationMode,
+    );
```
Comment (bradjc) [inline kernel/src/hil/eic.rs]: Is there a reason to separate out this function, and not just make it a part of `line_enable()`? Would a user want to enable an external interrupt but not configure it?

File: kernel/src/hil/eic.rs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+/// Enum for selecting syn/asyn mode
+pub enum SynchronizationMode {
+    Synchronous,
+    Asynchronous,
+}
+
+/// Interface for EIC.
+pub trait ExternalInterruptController {
+    /// Enables EIC module
+    fn enable(&self);
```
Comment (bradjc) [inline kernel/src/hil/eic.rs]: I think it would be best to remove the `enable()` and `disable()` functions, as I don't think we typically have them in Tock HILs. A user is unlikely to want to enable the controller but then not use it, or enable a specific interrupt but not enable the actual controller. The implementation of the HIL should automatically enable/disable the actual controller rather than requiring the user to manage that.

File: chips/sam4l/src/eic.rs [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+    }
+
+    fn disable(&self) {
+        pm::disable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(false);
+    }
+
+    fn line_enable(&self, line_num: usize) {
+        let regs: &EicRegisters = &*self.registers;
+        let mask: u32 = 1 << line_num;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: I would prefer doing it as follows to avoid `shift left overflows`  and remove some boiler plate:

```rustc
if line_num <= 8 {
   let interrupt_line = 1 << line_num;
   let regs: &EicRegisters = &*self.registers;
   regs.en.write(Interrupt::INT.val(interrupt_line))
} else {
  debug!("interrupt line: {} not supported!", line_num);
}
``

File: chips/sam4l/src/eic.rs [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+            6 => regs.en.write(Interrupt::INT.val(mask)),
+            7 => regs.en.write(Interrupt::INT.val(mask)),
+            8 => regs.en.write(Interrupt::INT.val(mask)),
+            _ => debug!("not supported!"),
+        }
+
+        self.line_enable_interrupt(line_num);
+    }
+
+    fn line_disable(&self, line_num: usize) {
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: same here, see comment on line enable

File: chips/sam4l/src/eic.rs [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+    }
+
+    fn line_configure(
+        &self,
+        line_num: usize,
+        int_mode: hil::eic::InterruptMode,
+        filter: hil::eic::FilterMode,
+        syn_mode: hil::eic::SynchronizationMode,
+    ) {
+        let mask: u32 = 1 << line_num;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: what happends if line_num > 8 here?

could we do `if line_num > 8 { return; }`?

File: chips/sam4l/src/eic.rs [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+            7 => regs.icr.write(Interrupt::INT.val(mask)),
+            8 => regs.icr.write(Interrupt::INT.val(mask)),
+            _ => debug!("not supported!"),
+        }
+    }
+
+    pub fn line_is_enabled(&self, line_num: usize) -> bool {
+        let regs: &EicRegisters = &*self.registers;
+        let mask: u32 = 1 << line_num;
+        return (mask & regs.ctrl.get()) != 0;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        (mask & regs.ctrl.get()) != 0
```

File: chips/sam4l/src/eic.rs [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+            ],
+        }
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        self.enabled.get()
+    }
+
+    pub fn set_client(&self, client: &'a hil::eic::Client, line_num: usize) {
+        self.callbacks[line_num].set(client);
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: prefer `get()` over direct-indexing to avoid potential array out-of-bounds access

```suggestion
        self.callbacks.get(line_num).map(|c| c.set(client));
```

File: chips/sam4l/src/eic.rs [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+        self.callbacks[line_num].map(|cb| {
+            cb.fired();
+        });
+    }
+
+    /// Clears the interrupt flag of line. Should be called after handling interrupt
+    /// Sets interrupt clear register
+    pub fn line_clear_interrupt(&self, line_num: usize) {
+        let regs: &EicRegisters = &*self.registers;
+        let mask: u32 = 1 << line_num;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: grumble, on potential shift overflows see previous comments

Comment (jiashuoz) [reply chips/sam4l/src/eic.rs]: line_clear_interrupt() gets called in handle_interrupt(), which gets called in chip.rs. The line_num parameter passed into line_clear_interrupt() function can only be from 1-8. Should I do overflow checking here as well?

File: chips/sam4l/src/eic.rs (hunk 6 above)
Comment (jiashuoz) [reply chips/sam4l/src/eic.rs]: Thanks!

File: kernel/src/hil/eic.rs (hunk 1 above)
Comment (jiashuoz) [reply kernel/src/hil/eic.rs]: This is a question I need to explore a bit more: to figure out why application developers might want to use EIC. Right now the reason seems like an external interrupt in async mode can wake up the system from the deep sleep mode, whereas a GPIO pin interrupt cannot. If that's the case, then the question goes to "why not just set it always in async mode?". I'll add these as comments.

Comment (bradjc) [reply kernel/src/hil/eic.rs]: Well that is likely specific to the SAM4L (or maybe some subset of microcontrollers), but the ability to configure interrupt sources at this low level might be more generic. In particular, I think it would be helpful to guide the developer that is trying to add Tock support to a new MCU, and who needs to map hardware peripherals to HIL interfaces. I'm less worried about users higher up the stack that are trying to develop for, say, imix, and need to know when to use EIC vs. GPIO.

File: kernel/src/hil/eic.rs (hunk 3 above)
Comment (niklasad1) [reply kernel/src/hil/eic.rs]: Yes, it eliminate to check value of `line_num` too for example if you define an enum or something with only the possible variants :)

Comment (ppannuto) [reply kernel/src/hil/eic.rs]: Passing comment: please also update this to `line_number`, we err on the side of verbosity and clarity in HIL interfaces

File: kernel/src/hil/eic.rs (hunk 4 above)
Comment (jiashuoz) [reply kernel/src/hil/eic.rs]: I've merged enable()/disable() and also configure() into line_enable() function. But now I'm getting a kernel panic "Debug buffer full". In debug.rs, it says I could change the output_buffer size; maybe this was the issue? Any hint would be great!

Comment (bradjc) [reply kernel/src/hil/eic.rs]: If you try to use `debug!()` too often then it can fill the internal buffer faster than it can actually send the messages over uart and that error is the result. You can either remove some uses of `debug!()` or change the buffer size in kernel/debug.rs.

File: chips/sam4l/src/eic.rs [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+use kernel::hil;
+
+/// Representation of an EIC line on the SAM4L.
+pub struct EicLine {
+    line_number: u32,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u8)]
+enum Line {
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: You are definitively on right path but please use `CamelCase` for enum variants,

I would prefer something like:
```rustc
enum Line {
    Nmi = 1
    Ext1 = 2,
    Ext2 = 4,
    ....
    Ext8 = 256,
}
```
Then we could cast it directly to an u32 and don't need to shift when configuring interrupt(s).

File: chips/sam4l/src/eic.rs [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+    unsafe { StaticRef::new(0x400F1000 as *const EicRegisters) };
+
+pub struct Eic<'a> {
+    registers: StaticRef<EicRegisters>,
+    enabled: Cell<bool>,
+    callbacks: [OptionalCell<&'a hil::eic::Client>; 9],
+}
+
+impl<'a> hil::eic::ExternalInterruptController for Eic<'a> {
+    type Line = EicLine;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: I think the type `EicLine` is needless and should be removed

```suggestion
    type Line = Line;
```

File: chips/sam4l/src/eic.rs [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+
+use crate::pm::{self, Clock, PBDClock};
+use core::cell::Cell;
+use kernel::common::cells::OptionalCell;
+use kernel::common::registers::{register_bitfields, ReadOnly, ReadWrite, WriteOnly};
+use kernel::common::StaticRef;
+use kernel::hil;
+
+/// Representation of an EIC line on the SAM4L.
+pub struct EicLine {
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: Remove, the `Line` type should be sufficient for this use-case

File: chips/sam4l/src/eic.rs [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+    LINE3 = 0x03, // EXT3
+    LINE4 = 0x04, // EXT4
+    LINE5 = 0x05, // EXT5
+    LINE6 = 0x06, // EXT6
+    LINE7 = 0x07, // EXT7
+    LINE8 = 0x08, // EXT8
+}
+
+/// Initialization of an EIC line.
+impl EicLine {
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: remove

File: chips/sam4l/src/eic.rs [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+    const fn new(line: Line) -> EicLine {
+        EicLine {
+            line_number: ((line as u8) & 0x0F) as u32,
+        }
+    }
+}
+
+pub static mut LINE_EIC0: EicLine = EicLine::new(Line::LINE0);
+pub static mut LINE_EIC1: EicLine = EicLine::new(Line::LINE1);
+pub static mut LINE_EIC2: EicLine = EicLine::new(Line::LINE2);
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: remove these too

File: chips/sam4l/src/eic.rs [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+        line: &Self::Line,
+        int_mode: hil::eic::InterruptMode,
+        filter_mode: hil::eic::FilterMode,
+        syn_mode: hil::eic::SynchronizationMode,
+    ) {
+        if !self.is_enabled() {
+            self.enable();
+        }
+
+        let interrupt_line: u32 = 1 << line.line_number;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        let interrupt_line = *line as u32;
```

File: chips/sam4l/src/eic.rs [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+];
+
+// Page 59 of SAM4L data sheet
+const EIC_BASE: StaticRef<EicRegisters> =
+    unsafe { StaticRef::new(0x400F1000 as *const EicRegisters) };
+
+pub struct Eic<'a> {
+    registers: StaticRef<EicRegisters>,
+    enabled: Cell<bool>,
+    callbacks: [OptionalCell<&'a hil::eic::Client>; 9],
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
    callbacks: [OptionalCell<&'a dyn hil::eic::Client>; 9],
```

File: chips/sam4l/src/eic.rs (hunk 12 above)
Comment (niklasad1) [reply chips/sam4l/src/eic.rs]: Alternatively, explicit matching on the enum and configure the bit accordingly.

File: chips/sam4l/src/eic.rs (hunk 14 above)
Comment (jiashuoz) [reply chips/sam4l/src/eic.rs]: I removed EicLine. One more question though: am I supposed to make Line public? 

pub enum Line {
    Nmi = 1
    Ext1 = 2,
    Ext2 = 4,
    ....
    Ext8 = 256,
}

Comment (niklasad1) [reply chips/sam4l/src/eic.rs]: yeah I think so, since it is assigned to the `public associated type` otherwise it will be leaked I guess :P

File: chips/sam4l/src/eic.rs [hunk 19]
```diff
@@ -0,0 +1,321 @@
+//! Implementation of the SAM4L External Interrupt Controller (EIC).
+//!
+//! Datasheet section "21. External Interrupt Controller (EIC)".
+//!
+//! The External Interrupt Controller (EIC) allows pins to be configured as external
+//! interrupts. Each external interrupt has its own interrupt request and can be individually
+//! interrupt_lineed. Each external interrupt can generate an interrupt on rising or falling edge, or
```
Comment (ppannuto) [inline chips/sam4l/src/eic.rs]:
```suggestion
//! `interrupt_line`ed. Each external interrupt can generate an interrupt on rising or falling edge, or
```

File: chips/sam4l/src/eic.rs [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+        pm::enable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(true);
+    }
+
+    pub fn disable(&self) {
+        pm::disable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(false);
+    }
+
+    pub fn set_interrupt_mode(&self, mode_bits: u8, interrupt_line: u32) {
```
Comment (ppannuto) [inline chips/sam4l/src/eic.rs]: Should this be `pub`? It looks like it's only called as a helper method by `line_configure`. The reason I ask is that here the function is taking an opaque set of `mode_bits` where `line_configure` takes a nice `InterruptMode` type. This is interface is fine if this is internal (non-`pub`), but if it's meant to be external facing, it should probably use the better type in the interface

File: chips/sam4l/src/eic.rs [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+        }
+
+        let interrupt_line: u32 = 1 << *line as u32;
+        let regs: &EicRegisters = &*self.registers;
+        regs.dis.write(Interrupt::INT.val(interrupt_line));
+        self.line_disable_interrupt(interrupt_line);
+    }
+}
+
+impl<'a> Eic<'a> {
```
Comment (ppannuto) [inline chips/sam4l/src/eic.rs]: I called attention to one specific instance below, but more generally, should all of the methods here be `pub`? For those that should be public interfaces, can you ensure that they are all documented (`/// comment`)

File: kernel/src/hil/eic.rs [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    LowLevel,
+}
+
+/// Enum for enabling/disabling filter
+pub enum FilterMode {
+    FilterEnable,
+    FilterDisable,
+}
+
+/// Enum for selecting syn/asyn mode
```
Comment (ppannuto) [inline kernel/src/hil/eic.rs]:
```suggestion
/// Enum for selecting synchronous or asynchronous mode.
```

File: kernel/src/hil/eic.rs [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+/// Enum for selecting which edge to trigger interrupts on.
+#[derive(Debug)]
+pub enum InterruptMode {
+    RisingEdge,
+    FallingEdge,
+    HighLevel,
+    LowLevel,
+}
+
+/// Enum for enabling/disabling filter
```
Comment (ppannuto) [inline kernel/src/hil/eic.rs]:
```suggestion
/// Enum for enabling or disabling spurious event filtering (i.e. de-bouncing control).
```

File: kernel/src/hil/eic.rs [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+pub enum FilterMode {
```
Comment (ppannuto) [inline kernel/src/hil/eic.rs]: As a more general question: I've not encountered a software configurable interrupt filter before, but that could easily be my ignorance of the world. Do we think this is a SAM4L-specific thing (i.e. shouldn't be included as part of the HIL), or have others seen this elsewhere?

File: kernel/src/hil/eic.rs [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+pub trait ExternalInterruptController {
+    type Line;
+
+    /// Enables external interrupt on line_num
+    /// In asychronous mode, all edge interrupts will be
+    /// interpreted as level interrupts and the filter is disabled.
+    fn line_enable(
+        &self,
+        line: &Self::Line,
+        int_mode: InterruptMode,
```
Comment (ppannuto) [inline kernel/src/hil/eic.rs]:
```suggestion
        interrupt_mode: InterruptMode,
```

File: kernel/src/hil/eic.rs [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+        filter: FilterMode,
+        syn_mode: SynchronizationMode,
```
Comment (ppannuto) [inline kernel/src/hil/eic.rs]:
```suggestion
        synchronization_mode: SynchronizationMode,
```

File: kernel/src/hil/eic.rs [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+    type Line;
+
+    /// Enables external interrupt on line_num
+    /// In asychronous mode, all edge interrupts will be
+    /// interpreted as level interrupts and the filter is disabled.
+    fn line_enable(
+        &self,
+        line: &Self::Line,
+        int_mode: InterruptMode,
+        filter: FilterMode,
```
Comment (ppannuto) [inline kernel/src/hil/eic.rs]:
```suggestion
        filter_mode: FilterMode,
```

Either `_mode` on all of these or none of these I think

File: kernel/src/hil/eic.rs [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+        syn_mode: SynchronizationMode,
+    );
+
+    /// Disables external interrupt on line_num
```
Comment (ppannuto) [inline kernel/src/hil/eic.rs]:
```suggestion
    /// Disables external interrupt on the given `line`.
```

File: chips/sam4l/src/eic.rs (hunk 20 above)
Comment (jiashuoz) [reply chips/sam4l/src/eic.rs]: This should be private. Thanks for catching that! I've fixed others as well. For those functions that return a bool type, I used them mostly for testing purposes and made them pub but the users wouldn't use these functions in any scenario; is it better to remove these functions or leave them be?

Comment (ppannuto) [reply chips/sam4l/src/eic.rs]: Hmm.., this is trickier -- we don't have the greatest modular testing story in Tock right now. There are a few places that use `#[cfg(test)]`, but I'll confess to ignorance of the details. I'm inclined not to leave test code lying around, especially with pub visibility, but I'll defer to others on the best way to handle this.

Comment (niklasad1) [reply chips/sam4l/src/eic.rs]: I think `#[cfg(test)]` doesn't work because it pulls in `std` however https://doc.rust-lang.org/unstable-book/language-features/custom-test-frameworks.html looks like an option or define our own feature flag for test builds

File: chips/sam4l/src/eic.rs [hunk 29]
```diff
@@ -0,0 +1,321 @@
...
+use crate::pm::{self, Clock, PBDClock};
+use core::cell::Cell;
+use kernel::common::cells::OptionalCell;
+use kernel::common::registers::{register_bitfields, ReadOnly, ReadWrite, WriteOnly};
+use kernel::common::StaticRef;
+use kernel::hil;
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u8)]
+pub enum Line {
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: docs please, i.e, with `///` including each enum variant

File: chips/sam4l/src/eic.rs [hunk 30]
```diff
@@ -0,0 +1,321 @@
...
+        let regs: &EicRegisters = &*self.registers;
+        regs.dis.write(Interrupt::INT.val(interrupt_line));
+        self.line_disable_interrupt(interrupt_line);
+    }
+}
+
+impl<'a> Eic<'a> {
+    pub fn line_configure(
+        &self,
+        interrupt_line: u32,
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]: I think you can change this to
```suggestion
        line: Line,
```

File: kernel/src/hil/eic.rs (hunk 24 above)
Comment (bradjc) [reply kernel/src/hil/eic.rs]: I can't think of having seen this before, and the SAM4L definitely took a kitchen sink approach. However, in general this _seems_ like a feature that could exist on any chip. I think it makes sense to add a HIL interface, and see if it ever gets used in the future. My two cents.

Comment (bradjc) [reply kernel/src/hil/eic.rs]: Oh wow I totally professored this. I now realize what you are referring to. Yeah...this seems like it would be very difficult or impossible to implement [fully] on other chips.

File: kernel/src/hil/eic.rs [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+    /// Enables external interrupt on the given 'line'
+    /// In asychronous mode, all edge interrupts will be
+    /// interpreted as level interrupts and the filter is disabled.
+    fn line_enable(
+        &self,
+        line: &Self::Line,
+        interrupt_mode: InterruptMode,
+        filter_mode: FilterMode,
+        synchronization_mode: SynchronizationMode,
+    );
```
Comment (bradjc) [inline kernel/src/hil/eic.rs]: Along the lines of @ppannuto above, maybe this function should only accept `line` and `interrupt_mode` to make it easier for other chips to implement this HIL interface. Also, I imagine users don't want to think about these details? If they do, we could do something like UART and have an advanced trait that exposes more functionality.

File: chips/sam4l/src/eic.rs [hunk 32]
```diff
@@ -0,0 +1,339 @@
...
+    filter: ReadWrite<u32, Interrupt::Register>,       // 0x20
+    test: ReadWrite<u32, Test::Register>,              // 0x24
+    asynchronous: ReadWrite<u32, Interrupt::Register>, // 0x28
+    _reserved0: ReadOnly<u32>,                         // 0x02c, skip
+    en: WriteOnly<u32, Interrupt::Register>,           // 0x30
+    dis: WriteOnly<u32, Interrupt::Register>,          // 0x34
+    ctrl: ReadOnly<u32, Interrupt::Register>,          // 0x38
+}
+
+// IER: Writing a one to this bit will set the corresponding bit in IMR.
```
Comment (ppannuto) [inline chips/sam4l/src/eic.rs]: Can you move all these comments into doc comments above their register (like https://github.com/tock/tock/blob/master/chips/nrf52/src/adc.rs#L10)?

It's a bit of a newer Tock convention, but a good one I think

File: chips/sam4l/src/eic.rs [hunk 33]
```diff
@@ -0,0 +1,339 @@
...
+//          1: The corresponding external interrupt is enabled.
+
+register_bitfields![
+    u32,
+    Interrupt [
+        INT OFFSET(0) NUMBITS(32) []
+    ],
+    // Test is not being used right now
+    Test [
+        //0: This bit disables external interrupt test mode.
```
Comment (ppannuto) [inline chips/sam4l/src/eic.rs]: Similarly, promote these to doc comments (`//` -> `///`)

File: chips/sam4l/src/eic.rs [hunk 34]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    /// Enables the propagation from the EIC to the interrupt controller of the external interrupt
+    /// on a specified line.
+    fn line_enable_interrupt(&self, line: &Line) {
+        let regs: &EicRegisters = &*self.registers;
+
+        regs.ier.write(Interrupt::INT.val(*line as u32));
+    }
+
```
Comment (ppannuto) [inline chips/sam4l/src/eic.rs]:
```suggestion
    /// Disables the propagation from the EIC to the interrupt controller of the external interrupt
    /// on a specified line.
```

File: chips/sam4l/src/eic.rs [hunk 35]
```diff
@@ -0,0 +1,339 @@
...
+        if !self.is_enabled() {
+            return;
+        }
+
+        let regs: &EicRegisters = &*self.registers;
+        regs.dis.write(Interrupt::INT.val(*line as u32));
+
+        self.line_disable_interrupt(line);
+
+        self.disable();
```
Comment (ppannuto) [inline chips/sam4l/src/eic.rs]: Hm.. actually, thinking about this a bit more, I think this line is a problem -- what happens if you have multiple lines enabled, and then one line chooses to disable? I think that will incorrectly turn off the entire EIC?

Ideally, this would remove the explicit `enable`/`disable` calls and use the [automated peripheral management](https://docs.tockos.org/kernel/common/peripherals/index.html) instead ([here's a SPI example](https://github.com/tock/tock/blob/master/chips/sam4l/src/spi.rs#L200)). The `after_peripheral_access` method can probably just check `regs.imr.read() == 0`

**Comment by jiashuoz** [comment]
As Pat pointed out, that was indeed a bug. Thank you! Will fix soon.

Josh Zhang
Master Student in Computer Science
Princeton Class of 2020
(206) 953-5880
________________________________

File: chips/sam4l/src/eic.rs (hunk 35 above)
Comment (jiashuoz) [reply chips/sam4l/src/eic.rs]: if regs.imr.read() == 0 && regs.ctrl.read() == 0 {
       self.disable();
}

Adding this check would fix this right? Is it necessary to use automated peripheral management in this case?

Comment (ppannuto) [reply chips/sam4l/src/eic.rs]: Maybe not "necessary", but architecturally it's probably a bit better. Right now this code tracks the same piece of state (is the EIC enabled) in two places: (1) in the explicit `enabled` Cell, and (2) in the EIC registers. That means that whenever code updates (2) you also have to remember to correctly update (1). With automated peripheral management, you only store the state in (2), which is architecturally cleaner and eliminates possible state synchronization bugs.

Comment (jiashuoz) [reply chips/sam4l/src/eic.rs]: For the most current commit, I'm simply reading the register values to determine when to enable/disable the clock. The bug is fixed. I removed the explicit enabled Cell. I tried using peripheral management method but I kind of think it's a bit not worth using it for EIC. I will look into it later.

File: chips/sam4l/src/eic.rs [hunk 36]
```diff
@@ -0,0 +1,347 @@
...
+
+pub struct Eic<'a> {
+    callbacks: [OptionalCell<&'a dyn hil::eic::Client>; 9],
+}
+
+impl<'a> hil::eic::ExternalInterruptController for Eic<'a> {
+    type Line = Line;
+
+    fn line_enable(&self, line: &Self::Line, interrupt_mode: hil::eic::InterruptMode) {
+        let regs = &EicRegisterManager::new(&self).registers;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        let regs = self.get_registers;
```

Comment (niklasad1) [reply chips/sam4l/src/eic.rs]: I guess that it is not that costly to create `PeripheralManager::new` but it is already implemented and would be better just to call the method instead of creating the new value and just use it for the registers.

File: chips/sam4l/src/eic.rs [hunk 37]
```diff
@@ -0,0 +1,347 @@
...
+            FilterMode::FilterEnable,
+            SynchronizationMode::Asynchronous,
+        );
+
+        // enables propagation from eic to nvic, sets imr register
+        regs.ier.write(Interrupt::INT.val(*line as u32));
+    }
+
+    fn line_disable(&self, line: &Self::Line) {
+        let regs = &EicRegisterManager::new(&self).registers;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        let regs = self.get_registers();
```

File: chips/sam4l/src/eic.rs [hunk 38]
```diff
@@ -0,0 +1,347 @@
...
+        }
+
+        match synchronization_mode {
+            SynchronizationMode::Synchronous => self.line_disable_asyn(line),
+            SynchronizationMode::Asynchronous => self.line_enable_asyn(line),
+        }
+    }
+
+    fn set_interrupt_mode(&self, mode_bits: u8, line: &Line) {
+        let regs = &EicRegisterManager::new(&self).registers;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        let regs = self.get_registers();
```

File: chips/sam4l/src/eic.rs [hunk 39]
```diff
@@ -0,0 +1,347 @@
...
+
+    /// Registers a client associated with a line.
+    pub fn set_client(&self, client: &'a hil::eic::Client, line: &Line) {
+        self.callbacks.get(*line as usize).map(|c| c.set(client));
+    }
+
+    /// Executes client function when an interrupt is triggered.
+    pub fn handle_interrupt(&self, line: &Line) {
+        // Clears interrupt bit and then handle interrupt
+        let regs = &EicRegisterManager::new(&self).registers;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        let regs = self.get_registers();
```

File: chips/sam4l/src/eic.rs [hunk 40]
```diff
@@ -0,0 +1,347 @@
...
+
+        self.callbacks[*line as usize].map(|cb| {
+            cb.fired();
+        });
+    }
+
+    /// Returns true is a line is enabled. This doesn't mean the interrupt is being
+    /// propagated through. Developers can use this function for testing.
+    pub fn line_is_enabled(&self, line: &Line) -> bool {
+        let regs = &EicRegisterManager::new(&self).registers;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        let regs = self.get_registers();
```

File: chips/sam4l/src/eic.rs [hunk 41]
```diff
@@ -0,0 +1,347 @@
...
+
+        ((*line as u32) & regs.ctrl.get()) != 0
+    }
+
+    /// Returns true if interrupt is being propagated from EIC to the interrupt controller of
+    /// the external interrupt on a specific line, false otherwise. Developers can use this
+    /// function for testing.
+    pub fn line_interrupt_is_enabled(&self, line: &Line) -> bool {
+        let regs = &EicRegisterManager::new(&self).registers;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        let regs = self.get_registers();
```

File: chips/sam4l/src/eic.rs [hunk 42]
```diff
@@ -0,0 +1,347 @@
...
+
+        ((*line as u32) & regs.imr.get()) != 0
+    }
+
+    /// Returns true if a line's interrupt is pending, false otherwise. Developers can use this
+    /// function for testing.
+    pub fn line_interrupt_pending(&self, line: &Line) -> bool {
+        let regs = &EicRegisterManager::new(&self).registers;
```
Comment (niklasad1) [inline chips/sam4l/src/eic.rs]:
```suggestion
        let regs = self.get_registers();
```

File: chips/sam4l/src/eic.rs (hunk 36 above)
Comment (jiashuoz) [reply chips/sam4l/src/eic.rs]: Yea, this is better! I'm fixing these right now. Thanks!
